pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go

```diff
@@ -0,0 +1,159 @@
+	serviceClient corev1.CoreV1Interface
+
+	queue           queue.Instance
+	serviceInformer cache.SharedInformer
+
+	clusterLocalHosts []string //hosts marked as cluster-local, which will not have serviceeexports generated
+}
+
+func NewServiceExportController(kubeClient kube.Client, clusterLocalHosts []string) (*ServiceExportController, error) {
+	serviceExportController := &ServiceExportController{
+		client:            kubeClient.MCSApis(),
+		serviceClient:     kubeClient.Kube().CoreV1(),
+		queue:             queue.NewQueue(time.Second),
+		clusterLocalHosts: clusterLocalHosts,
+	}
+
+	serviceExportController.serviceInformer = kubeClient.KubeInformer().Core().V1().Services().Informer()
+	serviceExportController.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			serviceExportController.queue.Push(func() error {
+				serviceObj, err := convertToService(obj)
+				if err != nil {
+					return err
+				}
+				return serviceExportController.HandleNewService(serviceObj)
+			})
+		},
+		DeleteFunc: func(obj interface{}) {
+			serviceExportController.queue.Push(func() error {
+				serviceObj, err := convertToService(obj)
+				if err != nil {
+					return err
+				}
+				return serviceExportController.HandleDeletedService(serviceObj)
+			})
+		},
+	})
+
+	return serviceExportController, nil
+}
+
+func (sc *ServiceExportController) Run(stopCh <-chan struct{}) {
+	log.Infof("Syncing existing services and serviceexports...")
+	sc.doInitialFullSync()
+	log.Infof("serviceexport sync complete")
+	cache.WaitForCacheSync(stopCh, sc.serviceInformer.HasSynced)
+	log.Infof("ServiceExport controller started")
+	go sc.queue.Run(stopCh)
+}
+
+func (sc *ServiceExportController) HandleNewService(obj *v1.Service) error {
+	if sc.isServiceClusterLocal(obj) {
+		return nil //don't don anything for marked clusterlocal services
+	}
+	return sc.createServiceExportIfNotPresent(obj)
+}
+
+func (sc *ServiceExportController) HandleDeletedService(obj *v1.Service) error {
+	return sc.deleteServiceExportIfPresent(obj)
+}
+
+func convertToService(obj interface{}) (*v1.Service, error) {
+	cm, ok := obj.(*v1.Service)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return nil, fmt.Errorf("couldn't get object from tombstone %#v", obj)
+		}
+		cm, ok = tombstone.Obj.(*v1.Service)
+		if !ok {
+			return nil, fmt.Errorf("tombstone contained object that is not a Service %#v", obj)
+		}
+	}
+	return cm, nil
+}
+
+func (sc *ServiceExportController) isServiceClusterLocal(service *v1.Service) bool {
+	for _, host := range sc.clusterLocalHosts {
+		hostComponents := strings.Split(host, ".")
+		//name match
+		if (hostComponents[0] == "*" || hostComponents[0] == service.Name) && (hostComponents[1] == "*" || hostComponents[1] == service.Namespace) {
+			return true
+		}
+	}
+	return false
+}
+
+func (sc *ServiceExportController) createServiceExportIfNotPresent(service *v1.Service) error {
```

Comment:
Who will consume the ServiceExports created here

Reply:
The short answer is anyone will be able to. I'm working in parallel with @nmittler on the implementation of the [Istio MCS RFC](https://docs.google.com/document/d/1K8hvQ83UcJ9a7U8oqXIefwr6pFJn-VBEi40Ak-fwQtk/edit#heading=h.xw1gqgyqs5b). I took "phase 2", and there is another issue open for "phase 1" (https://github.com/istio/istio/issues/29384)

The consumption of ServiceExports is out of scope for this PR.

Reply:
It is not clear how this works from the RFC, should wait for the draft implement.

As i read through the mcs-api desgin, it is not clear to me how this can solve multicluster topology issue. 

Can you elaborate on how the EndpointSlice create is used?  https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#importing-services

Reply:
@hzxuzhonghu 

The basic MCS flow is:

1. Something generates a `ServiceExport` for a service in a cluster
2. The MCS controller (Istio?) sees the `ServiceExport` and creates `ServiceImport` and `EndpointSlice`s in all of the *other* clusters in the clusterset (i.e. mesh).
3. In the other clusters, the `ServiceImport` and `EndpointSlice`s are used by K8s to program kube proxy to allow service calls to span clusters.

What this PR is attempting to accomplish is 1. This simply creates the `ServiceExport` resource ... it does nothing with `EndpointSlice`s. This will have no real impact on the system if there is no MCS controller performing the `ServiceImport`.